Mm 27913 search channel tip

TheDarkestDay opened the pull request:
#### Summary
Adds a new kind of toast - channel search hint toast. This toast is displayed only when user scrolls too far away in a channel history and no other toast is eligible to be in view. 

This toast simply notifies user about ability to use keyboard shortcut in order to quickly jump to search through channel content. 

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/15649

#### Related Pull Requests
Not this far (but I guess we have to introduce similar changes to the Desktop app?)

#### Screenshots
![image](https://user-images.githubusercontent.com/7871134/96996092-8ee5c080-152f-11eb-9f56-42e35a9a785d.png)

mattermod commented:
Hello @TheDarkestDay,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

hanzei commented on components/toast_wrapper/toast_wrapper.jsx:
```diff
@@ -254,6 +263,18 @@ class ToastWrapper extends React.PureComponent {
+    getSearchHintToastText = () => {
+        return (
+            <FormattedMessage
+                id='postlist.toast.searchHint'
```
Heads up that you need to add that translation to `i18n/en.json`. You can use `make i18n-extract` for that. See https://developers.mattermost.com/contribute/webapp/using-i18n-extract/ for more information.

TheDarkestDay replied:
Oh, sorry for this. Done.

esethna commented:
@TheDarkestDay this is awesome! Nice work! Adding @matthewbirtch for a UX review.

Regarding desktop, I don't think you need to make any changes on that repo specifically, you'll just need to target the desktop app from here and show the different keyboard shortcut if so. @nevyangelova can probably provide guidance there

esethna commented:
@jgilliam17 when testing this please make sure there are no issues with this toast conflicting with the new messages toast. And let's check for regressions with the new messages toast. This search tip toast should never appear if the new messages toast is visible

esethna commented:
@TheDarkestDay excellent work with this, let us know if you'd be interested in taking on some more challenging front-end tickets! Examples:

https://github.com/mattermost/mattermost-server/issues/15023
https://github.com/mattermost/mattermost-server/issues/15658

Also feel free to reach out on our https://community.mattermost.com/ server if you're looking for something else! I'm @eric.sethna

nevyangelova reviewed (changes requested):
Thanks @TheDarkestDay wonderful work 👏 code looks really clean and organised, but we would like our new PR's to be written in TypeScript as we have an active campaign to migrate the app to TypeScript atm. Would you mind adding types? 

Note for the team: We might need to follow this approach when adding tests ourselves as this is a valuable thing to consider - adding test-window and reacting to window interactions.

matthewbirtch reviewed (changes requested):
Thanks for the contribution @TheDarkestDay. It looks good, however there are some descrepencies with the design files. I'll do my best below to try and outline those clearly.

**Here's the comparison**
<img width="469" alt="Frame 4" src="https://user-images.githubusercontent.com/2040554/97052239-9f596380-154e-11eb-9817-ea0563fc23dc.png">
If you'd like here's a link to the [Figma artboard](https://www.figma.com/file/FBBQQs6WiboS8shpMwVo2l/MM-27913-Search-Tip-Center-Channel?node-id=213%3A86) for the comparison

**Summary:**
- The toast container should be 16px down from channel header - looks to be 10px
- Left padding in the toast container should be 12px - looks to be 10px
- The height of the toast should be 40px, looks to be 32px
- The font-size on the toast message should be 14px, it looks to be 13px
- The text color for the toast message should be `color: var(--center-channel-color);`
- The font-size for the shortcut keys should be 14px
- The shortcut key containers should be 24px in height
- It also looks like the shortcut key containers need a bit more vertically and horizontal padding (which might make up the difference in height mentioned above)
- The background on the shortcut key containers should be `background-color: var(--center-channel-color-08);
- The space between the shortcut keys should be 4px, it looks to be 5px
- It looks like the content inside the toast might be slightly off-center vertically
- The background color for the toast container should be using `background-color: var(--center-channel-bg);`
- Need to add some margin on the right of the last shortcut key to match the spacing between 'Try' and the first shortcut key 
- The color of the 'x' dismiss icon it too faint. It should be `fill: var(--center-channel-color-64);`
- The toast container needs a border on it: `color: var(--center-channel-color);`
- Overall, all the colors should be using theme variables above as it doesn't end up looking right in dark themes for example
- The box shadow on the toast container should be `box-shadow: 0px 6px 14px rgba(0, 0, 0, 0.12);`

TheDarkestDay commented:
> Thanks @TheDarkestDay wonderful work 👏 code looks really clean and organised, but we would like our new PR's to be written in TypeScript as we have an active campaign to migrate the app to TypeScript atm. Would you mind adding types?
> 
> Note for the team: We might need to follow this approach when adding tests ourselves as this is a valuable thing to consider - adding test-window and reacting to window interactions.

Done.

TheDarkestDay commented:
Good time of the day, everyone (morning, I suppose?)

Just a little heads-up about all of the recent changes:

1. Desktop-app specific shortcuts should be displayed properly
2. Hopefully, all of the @matthewbirtch remarks are fixed to some extent. But I might be wrong, of course. 

By the way, @matthewbirtch may I kindly ask your advice about proper color for hover state of "Dismiss" button? Thanks in advance.

matthewbirtch commented:
> By the way, @matthewbirtch may I kindly ask your advice about proper color for hover state of "Dismiss" button? Thanks in advance.

Apologies @TheDarkestDay. use the following:
default state: `var(--center-channel-color-56)`
hover state: `var(--center-channel-color-72)`

I'll do another pass on the review for everything else

matthewbirtch commented on components/search_shortcut/search_shortcut.scss:
```diff
@@ -0,0 +1,9 @@
+.search-shortcut {
+  .shortcut-key {
+    margin-right: 4px;
```
add
```
display: inline-block;
```

matthewbirtch commented on components/search_shortcut/search_shortcut.scss:
```diff
@@ -0,0 +1,9 @@
+.search-shortcut {
```
add this to the `search-shortcut` class 
```
display: inline-block;
margin: 0 4px;
```

matthewbirtch commented on components/toast/toast.scss:
```diff
@@ -75,6 +75,59 @@
+
+    .toast__message {
+        position: static;
+
```
add
```
vertical-align: middle;
```

matthewbirtch commented on components/toast/toast.scss:
```diff
@@ -75,6 +75,59 @@
+            background-color: var(--center-channel-bg);
+
+            svg {
+                fill: rgba(0,0,0,0.2);
```
change this fill to:
```
fill: var(--center-channel-color-72);
```

matthewbirtch commented on components/toast/toast.scss:
```diff
@@ -75,6 +75,59 @@
+        }
+
+        svg {
+            fill: var(--center-channel-color-64);
```
sorry, I gave you the wrong opacity value for this fill. it should be 56
```
fill: var(--center-channel-color-56);
```

matthewbirtch reviewed (changes requested):
Thanks for the updates @TheDarkestDay. I added a few minor css tweaks.

Beyond that, I just noticed that the toast doesn't seem to be quite centered horizontally withing the view. I just took a quick measurement - see below

![image](https://user-images.githubusercontent.com/2040554/97185493-7f0bed80-1776-11eb-8288-3ff91b183131.png)

TheDarkestDay commented:
Done with all of the @matthewbirtch fixes. Thanks a lot for pointing these out.

matthewbirtch commented on components/toast/toast.scss:
```diff
@@ -75,6 +75,62 @@
+    margin-right: 0;
+    padding-top: 8px;
+    padding-bottom: 8px;
+    border: 1px solid var(--center-channel-color) inset;
```
Apologies, I missed this one. This border doesn't seem to be applying. 'inset' doesn't seem to work. You can just go with:

```
border: solid 1px var(--center-channel-color-16);
```

TheDarkestDay replied:
Border appearance is fixed, thanks a lot.

matthewbirtch reviewed (approved):
Just missed the border - I added a comment to that effect. Approving this provided the border gets updated

esethna commented:
@amyblais note that we should not release this change unless the associated keyboard shortcut is updated.

See this ticket: https://github.com/mattermost/mattermost-server/issues/15639 @MathewtheCoder opened a [PR for webapp](https://github.com/mattermost/mattermost-webapp/pull/6611), although I'm not sure if he's still working on the issue. We'll also need to open a PR for desktop app. @TheDarkestDay let us know if you're interested in making the [desktop changes as described here](https://github.com/mattermost/mattermost-webapp/pull/6611#issuecomment-702983786)

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

jfrerich commented:
@esethna, would it be appropriate to place the `Do Not Merge/Awaiting PR` until the linked PR is merged?

esethna commented:
Merging blocked on: https://github.com/mattermost/mattermost-webapp/pull/6611

jgilliam17 commented:
Thank you @TheDarkestDay - apologize for a delay in getting to this PR
The search tip looks good and does not conflict with the new message toast.
Seeing few other issues

1. Search tip is overlaid over `viewing message history` toast
steps:
Scroll in channel to trigger search tip toast
Click on any one of the visible post time stamps
<img width="1055" alt="Screen Shot 2020-11-09 at 10 39 53 AM" src="https://user-images.githubusercontent.com/52937121/98563592-59540d80-2279-11eb-830b-57561e98a750.png">

2. Search tip does not show in mobile browser view when scrolling, but pops up when browser is enlarged
steps:
Shrink the window to mobile size view 
Scroll 3+ pages to try to trigger the search toast - observed no search toast as expected
Enlarge the browser window (get out of mobile view) - observe search toast pops up on the screen without scrolling

3. Similarly if you reversed the steps from point # 2.  - Search tip toast is not dismissed when user switches to mobile view
steps:
While in standard browser (not in mobile view) scroll in long thread to trigger the search toast, do not dismiss it
Shrink the window to mobile view - observe search tip is still displayed in the mobile view
![Screenshot 2020-11-09 115824](https://user-images.githubusercontent.com/52937121/98572208-5e1dbf00-2283-11eb-989a-27826cd8882b.png)

4. Large Search tip toast when RHS is open. 
This may be acceptable, since we are wrapping the text here. @matthewbirtch Thoughts?
<img width="1046" alt="Screen Shot 2020-11-09 at 11 42 52 AM" src="https://user-images.githubusercontent.com/52937121/98570109-0c743500-2281-11eb-9e59-b02e2a4839a3.png">

matthewbirtch commented:
> 3\. Large Search tip toast when RHS is open.
>     This may be acceptable, since we are wrapping the text here. @matthewbirtch Thoughts?

@jgilliam17 I agree, this should be updated. We should span as much of the width as possible - it look like we have a lot more room to work with to prevent it from wrapping on to 3 lines

TheDarkestDay commented:
Thanks @jgilliam17 for this feedback. 

So, in general I have no questions regarding most of the points except (2).

Correct me if I am wrong but I have been assuming all this time that from a user perspective we differentiate "mobile" and "desktop" view of Mattermost by dimensions of the user screen (width and height). With this being said, case number (2) looks valid - we had not seen this tooltip initially because app had been considered as in "mobile" view state, but then we shifted this view to "desktop" one by enlarging our browser window. And we started to see this tooltip afterwards. 

For sure, I might be wrong with that assumption or may be I didn't get your case right.

jgilliam17 commented:
@TheDarkestDay ah, I see what you mean.
Maybe @matthewbirtch can advise here 
I'll try to explain my use case - I was not expecting the scrolling in mobile size window to trigger the tip (it didn't - as expected), and after scrolling was stopped in mobile view and the browser window was enlarged, resizing being the last action, I no longer expected search tip to be triggered.

TheDarkestDay commented:
Greetings, everyone. 

So, I've got some time recently to refresh this one a bit and here is the summary.

@jgilliam17 - It looks as if I've fixed all of the points you mentioned except (2), which is about hint popping up on screen enlargement, because this one appears kinda non-trivial. I will probably try to dig a little bit this one further but here is how it is currently looking: in order to determine, whether we should show this "hint" or not, we track "scroll" events (because ultimately our decision depends on current scrollbar position and other related things). The tricky part here is that browser window resizing also triggers these "scroll" events and as far as I can see for now we don't have a lot of things at our disposal which would allow us to distinguish between these "forced" scrolls and scrolls driven by a user. So, it would be great IMO to assess severity of this issue and maybe even create some separate issue if it turned out to be something which requires more fundamental changes in scroll/resizing tracking behavior. 

I've introduced some refactoring changes into `toast_wrapper` in order to attempt to move towards several small, more focused toasts instead of one toast tailored for all of the possible use cases. That is why I will highly appreciate any technical review of these changes as well as careful testing of all of the existing toast-related use cases.

TheDarkestDay commented:
Just FYI, I would be happy to help with https://github.com/mattermost/mattermost-webapp/pull/6611 as well just to give some momentum to all this work. Just let me know, if we can simply hang over this ticket to me smoothly. Not exactly sure how this process works at Mattermost. 

Thanks in advance.

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Thank you @TheDarkestDay 
- Verified Search channel tip toast does not conflict with new message toast.
- Verified points 1, 3 and 4 above are fixed; Search tip no longer overlays history toast, dismisses when entering mobile webview and not wrapping the text when RHS is open.
I agree that point 2., where search tip pops up when exiting mobile webview, is an edge case and will leave it to @matthewbirtch to decide if this is acceptable behavior.
Thanks again 🙂

matthewbirtch commented:
>   I agree that point 2., where search tip pops up when exiting mobile webview, is an edge case and will leave it to @matthewbirtch to decide if this is acceptable behavior.

I agree this is an edge case, but it looks like it reappears even if I have dismissed it. If I have previously dismissed it, shouldn't it be gone? Steps to reproduce:
1. In desktop view, scroll until you get the search tip
2. Dismiss the search tip toast
3. Change to mobile webview size
4. Change back to desktop size
5. The toast reappears

Again, this is not normal behaviour for a user, but it makes me wonder if there is another issue with dismissing the toast?

TheDarkestDay commented:
> > I agree that point 2., where search tip pops up when exiting mobile webview, is an edge case and will leave it to @matthewbirtch to decide if this is acceptable behavior.
> 
> I agree this is an edge case, but it looks like it reappears even if I have dismissed it. If I have previously dismissed it, shouldn't it be gone? Steps to reproduce:
> 
> 1. In desktop view, scroll until you get the search tip
> 2. Dismiss the search tip toast
> 3. Change to mobile webview size
> 4. Change back to desktop size
> 5. The toast reappears
> 
> Again, this is not normal behaviour for a user, but it makes me wonder if there is another issue with dismissing the toast?

You are right, I've missed this thing a little. Going to fix it soon. Thank you.

TheDarkestDay commented:
@matthewbirtch @jgilliam17 

I've just pushed a fix for this last issue with toast appearing even after dismissal.

matthewbirtch commented:
> I've just pushed a fix for this last issue with toast appearing even after dismissal.

Fixed for me!

jgilliam17 reviewed (approved):
Thank you for your contribution @TheDarkestDay 
Tested, looks good to merge.
- Verified new channel search tip that appears after user scrolls through long list of posts. No conflicts with history or new message toast, dismisses when entering mobile webview and not wrapping the text when RHS is open.

hmhealey reviewed (approved):
Thanks for the PR, @TheDarkestDay! Looks good.

hanzei commented:
@esethna Per [this comment](https://github.com/mattermost/mattermost-webapp/pull/6896#issuecomment-716767141): Should this PR be merged now or do we have to wait?

hanzei commented:
/update-branch

esethna commented:
@TheDarkestDay open for thoughts, perhaps we should change the keyboard shortcut in the tip to CMD/CTRL + F so we can merge this without being blocked on the shortcut being implemented here: https://github.com/mattermost/mattermost-webapp/pull/6611

Alternatively, @TheDarkestDay feel free to provide some guidance or suggestions to @MathewtheCoder to help him finish the PR

@jgilliam17 was this tested on desktop, mac and windows? What shortcut is shown on desktop?

jgilliam17 commented:
> @jgilliam17 was this tested on desktop, mac and windows? What shortcut is shown on desktop?

This was tested on windows, macOS and Ubuntu, on both browser and desktop. If I remember correctly, desktop shortcut did not have `shift`,  it was ctrl  F and cmd F, respectively

TheDarkestDay commented:
> @TheDarkestDay open for thoughts, perhaps we should change the keyboard shortcut in the tip to CMD/CTRL + F on browser (same as desktop) so we can merge this without being blocked on the CMD/CTRL + Shift + F shortcut being implemented for browser in this PR: #6611
> 
> Alternatively, @TheDarkestDay feel free to provide some guidance or suggestions to @MathewtheCoder to help him finish the PR

At this point I would rather try helping pushing shortcut implementation PR forward than falling back here to a browser-provided search. I haven't seen the whole problem yet in that PR but for now it seems kinda bulky to merge this PR with some fallback implementation and then create yet another PR with originally intended shortcut. 

On the other hand - there are not so many efforts actually required to change the shortcut text.

Anyway - I will try to help with #6611 as soon as possible.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

hmhealey commented:
/update-branch

TheDarkestDay commented:
Hey, guys. As far as I got we've done with the search shortcut implementation (https://github.com/mattermost/mattermost-webapp/pull/7718)

Are we ready to start reanimating this one?

esethna commented:
@TheDarkestDay yes, let's get this merged! cc// @devinbinnie

esethna commented:
@TheDarkestDay I noticed a minor issue with this on our daily build server, the cursor doesn't change to the hand when hovering over the "x" button. Let us know if you're open to submitting a PR to address it. Nice work on it!
 
![image](https://user-images.githubusercontent.com/14333569/116320413-0dd0e580-a76d-11eb-950b-3d39bed95974.png)

TheDarkestDay commented:
@esethna sure, it seems not that challenging (hopefully). 

What is the best way to submit a hotfix like this? Do you guys have any recommended flow for this? Should I reference the same ticket?

esethna commented:
@TheDarkestDay I created a ticket here, you can reference it in a new PR. Thanks for your help! https://mattermost.atlassian.net/browse/MM-35301
